### ISSUE TITLE
Autostart according to root pattern detect Or slient failed for root detect

### ISSUE THREAD

**Issue by nanozuki** [neovim/nvim-lspconfig#1707]
### Language server

denols tsserver

### Requested feature

I wrote also deno projects and node projects, I want to find a way to detect projects to start the specific language server. Because denols and tsserver can't handle different environments.

For now, I found a workaround for this:

```lua
local util = require('lspconfig/util')
lspconfig.tsserver.setup({
  root_dir = util.root_pattern("tsconfig.json", "jsconfig.json", "package.json"),
})
lspconfig.denols.setup({
  root_dir = util.root_pattern('deno_root'),
})
```

For example, I enter a node project, the denols won't find root_dir, so it won't start. But I will get an error:

```lua
[lspconfig] Autostart for denols failed: matching root directory not detected.
```

I think this way is too hacking and will get an error message. Is there a better way to achieve this? Or can I make this error silent?

**Comment by spywhere** [comment]
Quite a coincident that I just face the same issue yesterday. This is what I did currently, please feel free to let me know if this looks good so I can made a PR 😄 

```diff
diff --git a/lua/lspconfig/configs.lua b/lua/lspconfig/configs.lua
index 32bc04d..6aa8ac5 100644
--- a/lua/lspconfig/configs.lua
+++ b/lua/lspconfig/configs.lua
@@ -237,7 +237,7 @@ function configs.__newindex(t, config_name, config_def)
       elseif config.single_file_support then
         local pseudo_root = util.path.dirname(buf_path)
         id = manager.add(pseudo_root, true)
-      else
+      elseif not config.optional then
         vim.notify(
           string.format('[lspconfig] Autostart for %s failed: matching root directory not detected.', config_name)
         )
```

Usage
```diff
 local util = require('lspconfig/util')
 lspconfig.tsserver.setup({
   root_dir = util.root_pattern("tsconfig.json", "jsconfig.json", "package.json"),
+  optional = true
 })
 lspconfig.denols.setup({
   root_dir = util.root_pattern('deno_root'),
+  optional = true
 })
```

**Comment by mjlbach** [comment]
It's just a message, I don't know why this bothers people so much. I'm just going to remove it https://github.com/neovim/nvim-lspconfig/pull/1711

I don't think we should add top level configuration in `setup` for a single notification.

**Comment by mjlbach** [comment]
Closed by #1711
